File: src/main/java/com/dyonovan/neotech/common/tileentity/machine/TileThermalBinder.java
package com.dyonovan.neotech.common.tileentity.machine;

import cofh.api.energy.EnergyStorage;
import cofh.api.energy.IEnergyReceiver;
import com.dyonovan.neotech.common.blocks.IExpellable;
import com.dyonovan.neotech.common.tileentity.BaseMachine;
import com.dyonovan.neotech.common.tileentity.InventoryTile;
import com.dyonovan.neotech.handlers.ItemHandler;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.server.gui.IUpdatePlayerListBox;
import net.minecraft.util.EnumFacing;

public class TileThermalBinder extends BaseMachine implements IEnergyReceiver, IUpdatePlayerListBox, IExpellable {

    public int currentProcessTime;
    public EnergyStorage energyRF;

    private static final int RF_TICK= 100;
    public static final int BASE_PROCESS_TIME = 200;
    public static final int INPUT_SLOT_1 = 0;
    public static final int INPUT_SLOT_2 = 1;
    public static final int INPUT_SLOT_3 = 2;
    public static final int INPUT_SLOT_4 = 3;
    public static final int MB_SLOT_INPUT = 4;
    public static final int MB_SLOT_OUTPUT = 5;

    public TileThermalBinder() {
        energyRF = new EnergyStorage(10000);
        currentProcessTime = 0;
        inventory = new InventoryTile(6);
    }

    @Override
    public void update() {
        if (!this.hasWorldObj() || getWorld().isRemote || currentProcessTime <= 0) return;
        //todo add into slots when mb inserted
        if (currentProcessTime > 0 && currentProcessTime <= BASE_PROCESS_TIME) {
            if (inventory.getStackInSlot(MB_SLOT_INPUT) == null) {
                currentProcessTime = 0;
                worldObj.markBlockForUpdate(pos);
                return;
            }
            if (energyRF.getEnergyStored() >= RF_TICK) {
                energyRF.extractEnergy(RF_TICK, false);
                ++currentProcessTime;
            }
        }
        if (currentProcessTime >= BASE_PROCESS_TIME) {
            int speed = 0;
            int efficiency = 0;
            int capacity = 0;
            boolean io = false;
            for (int i = 0; i < 4; i++) {
                if (getStackInSlot(i) != null) {
                    if (getStackInSlot(i).getItem() == ItemHandler.speedProcessor)
                        speed += getStackInSlot(i).stackSize;
                    else if (getStackInSlot(i).getItem() == ItemHandler.capRam)
                        capacity += getStackInSlot(i).stackSize;
                    else if (getStackInSlot(i).getItem() == ItemHandler.effFan)
                        efficiency += getStackInSlot(i).stackSize;
                    else if (getStackInSlot(i).getItem() == ItemHandler.ioPort)
                        io = true;

                    inventory.setStackInSlot(null, i);
                }
            }
            inventory.setStackInSlot(null, MB_SLOT_INPUT);
            writeToMB(speed, efficiency, capacity, io);
        }
    }

    public void writeToMB(int speed, int efficiency, int capacity, boolean io) {

        ItemStack newMB = new ItemStack(ItemHandler.upgradeMB, 1);
        NBTTagCompound tag = new NBTTagCompound();
        tag.setInteger("Speed", speed);
        tag.setInteger("Efficiency", efficiency);
        tag.setInteger("Capacity", capacity);
        tag.setBoolean("AutoOutput", io);
        newMB.setTagCompound(tag);
        inventory.setStackInSlot(newMB, MB_SLOT_OUTPUT);
    }

    public void mergeMB() {
        currentProcessTime = 1;
    }

    /*******************************************************************************************************************
     ************************************** Energy Functions ***********************************************************
     *******************************************************************************************************************/

    @Override
    public int receiveEnergy(EnumFacing from, int maxReceive, boolean simulate) {
        int actual = energyRF.receiveEnergy(maxReceive, simulate);
        if (actual > 0) this.getWorld().markBlockForUpdate(this.pos);
        return actual;
    }

    @Override
    public int getEnergyStored(EnumFacing from) {
        return energyRF.getEnergyStored();
    }

    @Override
    public int getMaxEnergyStored(EnumFacing from) {
        return energyRF.getMaxEnergyStored();
    }

    @Override
    public boolean canConnectEnergy(EnumFacing from) {
        return true;
    }

    @Override
    public void spawnActiveParticles(double x, double y, double z) {

    }

    /*******************************************************************************************************************
     ************************************** Inventory Functions ********************************************************
     *******************************************************************************************************************/

    @Override
    public int[] getSlotsForFace(EnumFacing side) {
        return new int[] {};
    }

    @Override
    public boolean canInsertItem(int index, ItemStack itemStackIn, EnumFacing direction) {
        return false;
    } //TODO

    @Override
    public boolean canExtractItem(int index, ItemStack stack, EnumFacing direction) {
        return false;
    }

    @Override
    public boolean isItemValidForSlot(int index, ItemStack stack) {
        switch (index) {

        }
        return true;
    }

    @Override
    public int getField(int id) {
        switch (id) {
            case 0:
                return currentProcessTime;
            default:
                return 0;
        }

    }

    @Override
    public void setField(int id, int value) {
        switch (id) {
            case 0:
                currentProcessTime = value;
        }
    }

    @Override
    public int getFieldCount() {
        return 1;
    }

    @Override
    public void clear() {
        for (int i = 0; i < inventory.getSizeInventory(); i++) {
            inventory.setStackInSlot(null, i);
        }
    }

    /*******************************************************************************************************************
     **************************************** Tile Functions ***********************************************************
     *******************************************************************************************************************/

    @Override
    public void readFromNBT(NBTTagCompound tag) {
        super.readFromNBT(tag);
        energyRF.readFromNBT(tag);
        inventory.readFromNBT(tag, this, ":MainInv");
        currentProcessTime = tag.getInteger("CurrentProcessTime");
    }

    @Override
    public void writeToNBT(NBTTagCompound tag) {
        super.writeToNBT(tag);
        energyRF.writeToNBT(tag);
        inventory.writeToNBT(tag, ":MainInv");
        tag.setInteger("CurrentProcessTime", currentProcessTime);
    }
}

File: src/main/scala/com/dyonovan/neotech/common/tiles/machines/TileThermalBinder.scala
package com.dyonovan.neotech.common.tiles.machines

import com.dyonovan.neotech.collections.UpgradeBoard
import com.dyonovan.neotech.common.tiles.{MachineProcessor, AbstractMachine}
import com.dyonovan.neotech.managers.ItemManager
import com.teambr.bookshelf.util.InventoryUtils
import net.minecraft.inventory.Container
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.FurnaceRecipes
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing
import net.minecraftforge.fml.relauncher.{Side, SideOnly}

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Dyonovan, Paul Davis <pauljoda>
 * @since August 21, 2015
 */
class TileThermalBinder extends MachineProcessor {


    final val INPUT1 = 0
    final val INPUT2 = 1
    final val INPUT3 = 2
    final val INPUT4 = 3
    final val MB_INPUT = 4
    final val MB_OUTPUT = 5

    var count = 0

    var isRunning = false
    var lastCount = 0

    /**
      * The initial size of the inventory
      *
      * @return
      */
    override def initialSize: Int = 6

    /**
      * Used to get how long it takes to cook things, you should check for upgrades at this point
      *
      * @return The time it takes in ticks to cook the current item
      */
    override def getCookTime : Int = {
        if(getUpgradeBoard != null && getUpgradeBoard.getProcessorCount > 0)
            (200 * getCount) - (getUpgradeBoard.getProcessorCount * 24)
        else
            200 * getCount
    }

    /**
      * Used to tell if this tile is able to process
      *
      * @return True if you are able to process
      */
    override def canProcess : Boolean = {
        if(isRunning && energy.getEnergyStored >= getEnergyCostPerTick) {
            if(getStackInSlot(MB_OUTPUT) == null && getStackInSlot(MB_INPUT) != null && getCount == lastCount)
                return true
        }
        isRunning = false
        false
    }

    /**
      * Get the output of the recipe
      *
      * @param stack The input
      * @return The output
      */
    override def getOutputForStack(stack: ItemStack): ItemStack = {
        if (stack != null && (stack.getItem == ItemManager.upgradeMBEmpty || stack.getItem == ItemManager.upgradeMBFull))
            new ItemStack(ItemManager.upgradeMBFull) //Just used to tell the Sided inventory that we have something, not actual output
        else
            null
    }

    def build(): Unit = {
        getStackInSlot(MB_INPUT).getItem match {
            case ItemManager.upgradeMBEmpty if hasSlotUpgrades =>
                val tag = writeToMB()
                val newMB = new ItemStack(ItemManager.upgradeMBFull)
                newMB.setTagCompound(tag)
                setInventorySlotContents(MB_OUTPUT, newMB)
                setInventorySlotContents(MB_INPUT, null)
            case ItemManager.upgradeMBFull if !hasSlotUpgrades =>
                val mb = UpgradeBoard.getBoardFromStack(getStackInSlot(MB_INPUT))
                var slot = 0
                if (mb.hasControl) {
                    setInventorySlotContents(slot, new ItemStack(ItemManager.upgradeControl, 1))
                    slot += 1
                }
                if (mb.hasExpansion) {
                    setInventorySlotContents(slot, new ItemStack(ItemManager.upgradeExpansion, 1))
                    slot += 1
                }
                if (mb.getHardDriveCount > 0) {
                    setInventorySlotContents(slot, new ItemStack(ItemManager.upgradeHardDrive, mb.getHardDriveCount))
                    slot += 1
                }
                if (mb.getProcessorCount > 0) {
                    setInventorySlotContents(slot, new ItemStack(ItemManager.upgradeProcessor, mb.getProcessorCount))
                    slot += 1
                }
                setInventorySlotContents(MB_INPUT, null)
                setInventorySlotContents(MB_OUTPUT, new ItemStack(ItemManager.upgradeMBEmpty, 1))
            case _ =>
        }
    }

    override def doWork(): Unit = {
        if(getStackInSlot(MB_INPUT) == null) {
            reset()
        }
        if (this.values.burnTime > 0) {
            this.values.burnTime = values.burnTime - 1
            val actual = energy.extractEnergy(count * 10, true)
            if (actual < count * 10)
                reset()
            else
                energy.extractEnergy(actual, false)
            worldObj.markBlockForUpdate(pos)
        } else if (this.values.currentItemBurnTime > 0 && this.values.burnTime <= 0) {
            build()
            reset()
        }
    }

    def reset(): Unit = {
        values.burnTime = 0
        values.currentItemBurnTime = 0
        count = 0
        worldObj.markBlockForUpdate(pos)
    }

    def getCount: Int = {
        if (getStackInSlot(MB_INPUT) != null) {
            getStackInSlot(MB_INPUT).getItem match {
                case ItemManager.upgradeMBEmpty if hasSlotUpgrades =>
                    var count = 0
                    for (i <- 0 to 3) {
                        if (getStackInSlot(i) != null)
                            count += getStackInSlot(i).stackSize
                    }
                    return count
                case ItemManager.upgradeMBFull if !hasSlotUpgrades =>
                    val mb = UpgradeBoard.getBoardFromStack(getStackInSlot(MB_INPUT))
                    if (mb != null) {
                        var count = mb.getHardDriveCount + mb.getProcessorCount
                        if (mb.hasControl) count += 1
                        if (mb.hasExpansion) count += 1
                        return count
                    }
                case _ =>
            }
        }
        0
    }

    private def writeToMB(): NBTTagCompound = {
        val tag = new NBTTagCompound
        for (i <- 0 to 3) {
            if (getStackInSlot(i) != null) {
                getStackInSlot(i).getItem match {
                    case item: ItemManager.upgradeControl.type =>
                        if (!tag.getBoolean("Control")) {
                            tag.setBoolean("Control", true)
                            setInventorySlotContents(i, null)
                        }
                    case item: ItemManager.upgradeExpansion.type =>
                        if (!tag.getBoolean("Expansion")) {
                            tag.setBoolean("Expansion", true)
                            setInventorySlotContents(i, null)
                        }
                    case item: ItemManager.upgradeHardDrive.type =>
                        val countTag = tag.getInteger("HardDrive")
                        if (countTag + getStackInSlot(i).stackSize <= 8) {
                            tag.setInteger("HardDrive", countTag + getStackInSlot(i).stackSize)
                            setInventorySlotContents(i, null)
                        } else if (countTag + getStackInSlot(i).stackSize > 8) {
                            tag.setInteger("HardDrive", 8)
                            getStackInSlot(i).stackSize -= 8 - countTag
                        }
                    case item: ItemManager.upgradeProcessor.type =>
                        val countTag = tag.getInteger("Processor")
                        if (countTag + getStackInSlot(i).stackSize <= 8) {
                            tag.setInteger("Processor", countTag + getStackInSlot(i).stackSize)
                            setInventorySlotContents(i, null)
                        } else if (countTag + getStackInSlot(i).stackSize > 8) {
                            tag.setInteger("Processor", 8)
                            getStackInSlot(i).stackSize -= 8 - countTag
                        }
                }
            }
        }
        tag
    }

    override def getOutputSlots : Array[Int] = Array(MB_OUTPUT)

    override def getInputSlots : Array[Int] = Array(INPUT1, INPUT2, INPUT3, INPUT4, MB_INPUT)

    def hasSlotUpgrades: Boolean = {
        for (i <- 0 to 3) {
            if (getStackInSlot(i) != null) return true
        }
        false
    }

    @SideOnly(Side.CLIENT) override def getCookProgressScaled(scaleVal: Int): Int =
        (((this.values.currentItemBurnTime - this.values.burnTime) * scaleVal) / Math.max(20 * 20 * count, 0.001)).toInt

    override def spawnActiveParticles(x: Double, y: Double, z: Double): Unit = {}

    /**
     * Get the output of the recipe
 *
     * @param stack The input
     * @return The output
     */
    override def recipe(stack: ItemStack): ItemStack = null

    /**
     * Used to output the redstone single from this structure
     *
     * Use a range from 0 - 16.
     *
     * 0 Usually means that there is nothing in the tile, so take that for lowest level. Like the generator has no energy while
     * 16 is usually the flip side of that. Output 16 when it is totally full and not less
     *
     * @return int range 0 - 16
     */
    override def getRedstoneOutput: Int = InventoryUtils.calcRedstoneFromInventory(this)

    override def getSlotsForFace(side: EnumFacing): Array[Int] = {
        side match {
            case EnumFacing.DOWN => Array[Int](MB_OUTPUT)
            case _ => Array[Int](INPUT1, INPUT2, INPUT3, INPUT4, MB_INPUT, MB_OUTPUT)
        }
    }

    override def isItemValidForSlot(slot: Int, stack: ItemStack): Boolean = {
        val item = stack.getItem
        if (slot >= 0 && slot <= 3)
            item.isInstanceOf[ItemManager.upgradeControl.type] || item.isInstanceOf[ItemManager.upgradeExpansion.type] ||
                    item.isInstanceOf[ItemManager.upgradeHardDrive.type] || item.isInstanceOf[ItemManager.upgradeProcessor.type]
        else if (slot == 4) item.isInstanceOf[ItemManager.upgradeMBEmpty.type] || item.isInstanceOf[ItemManager.upgradeMBFull.type]
        else false
    }

    override def canExtractItem(index: Int, stack: ItemStack, direction: EnumFacing): Boolean = index == 5

    override def canInsertItem(slot: Int, itemStackIn: ItemStack, direction: EnumFacing): Boolean = {
        isItemValidForSlot(slot, itemStackIn)
    }

    override def writeToNBT(tag: NBTTagCompound): Unit = {
        super.writeToNBT(tag)
        tag.setInteger("Count", count)
    }

    override def readFromNBT(tag: NBTTagCompound): Unit = {
        super.readFromNBT(tag)
        count = tag.getInteger("Count")
    }
}
